implementation/src/test/java/io/smallrye/mutiny/operators/UniCacheTest.java

```diff
@@ -273,4 +354,33 @@ public void onFailure(Throwable ignored) {
         uni.subscribe().withSubscriber(subscriber);
     }
 
+    @Test
+    public void testThatCacheInvalidationCausesResubscriptionAndCannotRace() {
+        AtomicInteger cacheValidationCount = new AtomicInteger(0);
+        BiPredicate<Integer, Throwable> validator = (i, f) -> cacheValidationCount.incrementAndGet() % 2 == 0;
+        AtomicInteger sourceSubscriptionCount = new AtomicInteger(0);
+        Uni<Integer> uni = Uni.createFrom().deferred(() -> Uni.createFrom().item(sourceSubscriptionCount.incrementAndGet()))
+                .cacheItemOrFailure().whilst(validator);
+
+        // Every second subscription causes a re-subscription to the upstream
+        // This must not cause race conditions
+        List<UniAssertSubscriber<Integer>> results = Collections.synchronizedList(new LinkedList<>());
+        Executor parallel = Executors.newFixedThreadPool(2);
+        for (int i = 0; i < 2000; i++) {
+            Runnable subscribe = () -> {
+                UniAssertSubscriber<Integer> subscriber = new UniAssertSubscriber<>(false);
+                uni.subscribe().withSubscriber(subscriber);
+                results.add(subscriber);
+            };
+            race(subscribe, subscribe, parallel);
+        }
+
+        // there are two subscriptions in each of the 2000 loop iterations
+        // every second of them the cache is invalidated an the UniCaches subscription renewed
+        // because of the `incrementAndGet()` the final counter state is one above the iteration count
+        assertThat(sourceSubscriptionCount.get()).isEqualTo(2001);
+        // cache validation count is (subscription count (2x2000) - 1) since the initial subscription will not be validated
+        assertThat(cacheValidationCount.get()).isEqualTo(3999);
+    }
+
```

Comment:
Which Uni producers are worth further cache invalidation tests?

Reply:
I would also test with immediate failure / item (but that is easy).
Testing with Uni creation using supplier (Called on every subscription) would be useful as they must not be called concurrently.